RFC.md

```diff
@@ -0,0 +1,413 @@
+        input_pos: Optional[Tensor] = None,
+    ) -> [Tensor, List[Tensor]]:
+        ...
+
+		# Move max seq length check from attention to here
+        if seq_len > self.max_seq_len:
+            raise ValueError("Too long")
+
+        ...
+
+		hidden = []
+        for i, layer in enumerate(self.layers):
+	        if i in self.hidden_idx:
+		        hidden.append(h)
+            h = layer(
+                h,
+                mask=mask,
+                encoder_input=encoder_input,
+                encoder_mask=encoder_mask,
+                input_pos=input_pos,
+            )
+
+        # shape: [b, s, d]
+        h = self.norm(h)
+
+        output = self.output(h).float()
+        # this unifies the output with ViT but breaks recipes
+        # Alternative:
+        # output = output if not hidden else hidden + [output]
+        return output, hidden
+```
+@SalmanMohammadi does this design address all your needs from #1017
+### TransformerSelfAttentionLayer
+TransformerDecoderLayer only highlighting the changes, current module [here](https://github.com/pytorch/torchtune/blob/0057fe7cf83e14f0b62538a8d4d20719e0a88639/torchtune/modules/transformer.py#L15). Discussion in the comments.
+
+```python
+class TransformerSelfAttentionLayer(nn.Module):
+    def __init__(
+        self,
+        attn: GroupedQueryAttention, # Updated Attn module
+        mlp: nn.Module,
+        *,
+        # sa_norm renamed since it's no longer only self attention
+        # this requires repo wide changes for sa_norm -> attn_norm
+        attn_norm: Optional[nn.Module] = None,
+        mlp_norm: Optional[nn.Module] = None,
+        # New scale modules to allow scaling/gating attn output
+        attn_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
+    ) -> None:
+        super().__init__()
+        self.attn = attn
+        self.mlp = mlp
+        self.attn_norm = attn_norm or nn.Identity()
+        self.mlp_norm = mlp_norm or nn.Identity()
+        self.attn_scale = attn_scale or nn.Identity()
+        self.mlp_scale = mlp_scale or nn.Identity()
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs, # encoder_input and encoder_mask is not used for this module
+    ) -> Tensor:
+	    # Identical to TransformerDecoderLayer with addition of *_scale
+        attn_out = self.attn(self.attn_norm(x), mask=mask, input_pos=input_pos)
+
+        h = self.attn_scale(attn_out) + x
+
+        mlp_out = self.mlp(self.mlp_norm(h))
+
+        out = h + self.mlp_scale(mlp_out)
+        return out
+```
+
+### TransformerCrossAttentionLayer
+New layer. Discussion in the comments.
+```python
+class TransformerCrossAttentionLayer(nn.Module):
+    def __init__(
+		...
+    ) -> None:
+        super().__init__()
+        assert (
+            attn.pos_embeddings is None
+        ), "Positions are not computed for encoder inputs"
```

Comment:
Is there a nicer way to do this? This will limit the flexibility of this module a bit, right?

Reply:
Could you clarify more here? The flexibility of the attention blocks it can be used with? We could remove the check all together but thought it was better to have it until we didn't need it anymore.